### PR TITLE
[SPARK-33448][SQL] Support CACHE/UNCACHE TABLE commands for v2 tables

### PR DESCRIPTION
**Pull request by imback82**
### What changes were proposed in this pull request?

This PR proposes to support `CHACHE/UNCACHE TABLE` commands for v2 tables.

In addtion, this PR proposes to migrate `CACHE/UNCACHE TABLE` to use `UnresolvedTableOrView` to resolve the table identifier. This allows consistent resolution rules (temp view first, etc.) to be applied for both v1/v2 commands. More info about the consistent resolution rule proposal can be found in [JIRA](https://issues.apache.org/jira/browse/SPARK-29900) or [proposal doc](https://docs.google.com/document/d/1hvLjGA8y_W_hhilpngXVub1Ebv8RsMap986nENCFnrg/edit?usp=sharing).

### Why are the changes needed?

To support `CACHE/UNCACHE TABLE` commands for v2 tables.

Note that `CACHE/UNCACHE TABLE` for v1 tables/views go through `SparkSession.table` to resolve identifier, which resolves temp views first, so there is no change in the behavior by moving to the new framework.

### Does this PR introduce _any_ user-facing change?

Yes. Now the user can run `CACHE/UNCACHE TABLE` commands on v2 tables.

### How was this patch tested?

Added/updated existing tests.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala [hunk 1]
```diff
@@ -570,12 +567,9 @@ class ResolveSessionCatalog(
             "SHOW VIEWS, only SessionCatalog supports this command.")
       }
 
-    case ShowTableProperties(ResolvedV1TableIdentifier(ident), propertyKey) =>
+    case ShowTableProperties(ResolvedV1TableOrViewIdentifier(ident), propertyKey) =>
```
Comment (imback82) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:570]: A refactoring not related to this PR.

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala [hunk 2]
```diff
@@ -441,19 +441,16 @@ class ResolveSessionCatalog(
...
-        // Temp view doesn't belong to any catalog and we shouldn't resolve catalog in the name.
-        tbl
-      } else {
-        parseTempViewOrV1Table(tbl, "CACHE TABLE")
-      }
-      CacheTableCommand(name.asTableIdentifier, plan, isLazy, options)
+    // CACHE TABLE ... AS SELECT creates a temp view with the input query.
+    // Thus, use the identifier in UnresolvedTableOrView directly,
+    case CacheTable(u: UnresolvedTableOrView, plan, isLazy, options) if plan.isDefined =>
+      CacheTableCommand(u.multipartIdentifier.asTableIdentifier, plan, isLazy, options)
```
Comment (imback82) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:447]: @cloud-fan Please let me know what you think about having `UnresolvedTableOrView` here to eagerly use the identifier if `plan` is defined. Another approach is to have a separate rule to handle `CacheTable(u: UnresolvedTableOrView, ...)`.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/CachedTableSuite.scala [hunk 3]
```diff
@@ -113,7 +113,7 @@ class CachedTableSuite extends QueryTest with SQLTestUtils with TestHiveSingleto
     e = intercept[AnalysisException] {
       sql("UNCACHE TABLE nonexistentTable")
     }.getMessage
-    assert(e.contains(s"$expectedErrorMsg default.nonexistentTable"))
+    assert(e.contains(s"$expectedErrorMsg nonexistentTable"))
```
Comment (imback82) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/CachedTableSuite.scala:116]: This is now consistent with Line 108 above.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala [hunk 4]
```diff
@@ -416,9 +416,9 @@ case class DropTable(
 }
 
 /**
- * The logical plan for handling non-existing table for DROP TABLE command.
+ * The logical plan for no-op command handling non-existing table.
  */
-case class NoopDropTable(multipartIdentifier: Seq[String]) extends Command
+case class NoopCommand(multipartIdentifier: Seq[String]) extends Command
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala:421]: We should probably add a `commandName: String` property, which can be `DROP TABLE`, `REFRESH TABLE`, etc., so that we can see the original commannd name from the EXPLAIN result.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala [hunk 5]
```diff
@@ -289,6 +289,12 @@ class DataSourceV2Strategy(session: SparkSession) extends Strategy with Predicat
     case ShowCreateTable(_: ResolvedTable, _) =>
       throw new AnalysisException("SHOW CREATE TABLE is not supported for v2 tables.")
 
+    case CacheTable(_: ResolvedTable, _, _, _) =>
+      throw new AnalysisException("CACHE TABLE is not supported for v2 tables.")
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala:293]: We don't need new v2 APIs to support it. This command touches `CacheManager` which is Spark internal.

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala [hunk 6]
```diff
@@ -441,19 +441,16 @@ class ResolveSessionCatalog(
...
-    case CacheTableStatement(tbl, plan, isLazy, options) =>
-      val name = if (plan.isDefined) {
-        // CACHE TABLE ... AS SELECT creates a temp view with the input query.
-        // Temp view doesn't belong to any catalog and we shouldn't resolve catalog in the name.
-        tbl
-      } else {
-        parseTempViewOrV1Table(tbl, "CACHE TABLE")
-      }
-      CacheTableCommand(name.asTableIdentifier, plan, isLazy, options)
+    // CACHE TABLE ... AS SELECT creates a temp view with the input query.
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:444]: What's the behavior of it if the temp view already exists? overwrite?

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:447]: This is more like CTAS and the `table` should be just `Seq[String]` not `LogicalPlan`.

How about we have both `CacheTable(table: LogicalPlan, ...)` and `CacheTabeAsSelect(tempViewName: String, ...)`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala (hunk 5 above)
Comment (imback82) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala:293]: Ah OK. An existing bug, I guess? (it only supported temp view / v1 tables).

Does it make sense to match `case CacheTable(_: ResolvedTable, _, _, _) ` in `ResolveSessionCatalog` or should we match it in `DataSourceV2Strategy` with a new `CacheTableExec` similar to `CacheTableCommand`?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala:293]: We can add a `CacheTableExec` as a v2 version of `CacheTableCommand`.

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala (hunk 6 above)
Comment (imback82) [reply sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:444]: It would fail with:
```
org.apache.spark.sql.catalyst.analysis.TempTableAlreadyExistsException: Temporary view 't' already exists;
```

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala (hunk 2 above)
Comment (imback82) [reply sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:447]: Good idea

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala (hunk 4 above)
Comment (imback82) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala]: Added `commandName`. Now `EXPLAIN EXTENDED DROP TABLE` looks like the following:
```
|== Parsed Logical Plan ==
'DropTable true, false
+- 'UnresolvedTableOrView [testcat, ns1, ns2, tbl], true

== Analyzed Logical Plan ==
NoopCommand DROP TABLE, [testcat, ns1, ns2, tbl]

== Optimized Logical Plan ==
NoopCommand DROP TABLE, [testcat, ns1, ns2, tbl]

== Physical Plan ==
LocalTableScan <empty>
```

Btw, do we want to introduce `NoopCommandExec` for physical plan as well?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala (hunk 5 above)
Comment (imback82) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala]: Added v2 version.

**Comment by viirya** [comment]
cc @sunchao

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 7]
```diff
@@ -3323,7 +3323,7 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with SQLConfHelper with Logg
   }
 
   /**
-   * Create a [[CacheTableStatement]].
+   * Create a [[CacheTable]] pr [[CacheTableAsSelect]].
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:3326]: pr -> or?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CacheTableExec.scala [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+ */
+case class UncacheTableExec(
+    session: SparkSession,
+    catalog: TableCatalog,
+    table: Table,
+    ident: Identifier) extends V2CommandExec {
+  override def run(): Seq[InternalRow] = {
+    val v2Relation = DataSourceV2Relation.create(table, Some(catalog), Some(ident))
+    val df = Dataset.ofRows(session, v2Relation)
+    // Cascade should be true unless a temporary view is uncached.
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CacheTableExec.scala:73]: This comment reads confusing me. Is temp view uncaching also done by `UncacheTableExec`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 7 above)
Comment (imback82) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:3326]: Thanks, fixed.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CacheTableExec.scala (hunk 8 above)
Comment (imback82) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CacheTableExec.scala]: OK, I removed this comment.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 9]
```diff
@@ -3343,14 +3343,21 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with SQLConfHelper with Logg
         "the table name in CACHE TABLE AS SELECT", ctx)
     }
     val options = Option(ctx.options).map(visitPropertyKeyValues).getOrElse(Map.empty)
-    CacheTableStatement(tableName, query, ctx.LAZY != null, options)
+    val isLazy = ctx.LAZY != null
+    if (query.isDefined) {
+      CacheTableAsSelect(tableName.head, query.get, isLazy, options)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:3348]: nit: now we can move the if `tableName.length > 1` check into this branch.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CacheTableExec.scala [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+import org.apache.spark.sql.catalyst.expressions.Attribute
+import org.apache.spark.sql.connector.catalog.{Identifier, Table, TableCatalog}
+import org.apache.spark.sql.execution.CacheTableUtils
+import org.apache.spark.storage.StorageLevel
+
+/**
+ * Physical plan node for caching a table.
+ */
+case class CacheTableExec(
+    session: SparkSession,
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CacheTableExec.scala:31]: we can pass `CacheManager` to be more specific.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CacheTableExec.scala [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+  }
+
+  override def output: Seq[Attribute] = Seq.empty
+}
+
+/**
+ * Physical plan node for uncaching a table.
+ */
+case class UncacheTableExec(
+    session: SparkSession,
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CacheTableExec.scala:66]: ditto

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CacheTableExec.scala [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+    catalog: TableCatalog,
+    table: Table,
+    ident: Identifier) extends V2CommandExec {
+  override def run(): Seq[InternalRow] = {
+    val v2Relation = DataSourceV2Relation.create(table, Some(catalog), Some(ident))
+    val df = Dataset.ofRows(session, v2Relation)
+    session.sharedState.cacheManager.uncacheQuery(df, cascade = true)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CacheTableExec.scala:73]: we can call the `uncacheQuery` method that takes `LogicalPlan`

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CacheTableExec.scala:73]: we can call the `uncacheQuery` method that takes `LogicalPlan`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CacheTableExec.scala (hunk 10 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CacheTableExec.scala:31]: We don't need this parameter, we can access `SparkPlan.sqlContext.sparkSession`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CacheTableExec.scala [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+  override def run(): Seq[InternalRow] = {
+    import org.apache.spark.sql.connector.catalog.CatalogV2Implicits.IdentifierHelper
+
+    val v2Relation = DataSourceV2Relation.create(table, Some(catalog), Some(ident))
+    val df = Dataset.ofRows(session, v2Relation)
+    val tableName = Some(ident.quoted)
+    val optStorageLevel = CacheTableUtils.getStorageLevel(options)
+    if (optStorageLevel.nonEmpty) {
+      session.sharedState.cacheManager.cacheQuery(
+        df, tableName, StorageLevel.fromString(optStorageLevel.get))
```
Comment (sunchao) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CacheTableExec.scala:46]: should we include the catalog name in `tableName`? otherwise different catalogs may have tables with the same name?

Comment (imback82) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CacheTableExec.scala:46]: `tableName` is used only for display purpose (e.g., `InMemoryTableScanExec`). The `cachedData` is matched by the logical plan, so I think it should be OK.

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala [hunk 14]
```diff
@@ -710,6 +712,7 @@ class DataSourceV2SQLSuite
 
   test("DropTable: basic") {
     val tableName = "testcat.ns1.ns2.tbl"
+    sql(s"EXPLAIN EXTENDED DROP TABLE IF EXISTS $tableName").show(false)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:715]: is it for debug?

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala [hunk 15]
```diff
@@ -2018,28 +2021,32 @@ class DataSourceV2SQLSuite
     }
   }
 
-  test("CACHE TABLE") {
+  test("CACHE/UNCACHE TABLE") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:2024]: shall we also test CACHE TABLE AS SELECT? with temp view exists and not.

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala (hunk 14 above)
Comment (imback82) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala]: Oops. removed.

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala (hunk 15 above)
Comment (imback82) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:2023]: Added a test with an existing temp view to `CachedTableSuite`.

**Review by viirya — APPROVED** [review]

**Comment by viirya** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala [hunk 16]
```diff
@@ -679,3 +679,30 @@ case class TruncateTable(
     partitionSpec: Option[TablePartitionSpec]) extends Command {
   override def children: Seq[LogicalPlan] = child :: Nil
 }
+
+/**
+ * The logical plan of the CACHE TABLE command.
+ */
+case class CacheTable(
+    child: LogicalPlan,
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala:687]: After more thought, I think CACHE TABLE is not a DDL command that needs to interact with catalogs, and it doesn't need a v2 version.

The current problem is that `CacheTableCommand` only takes v1 table identifier and can't cache v2 tables with n part name. Maybe we can fix `CacheTableCommand` directly?

Comment (imback82) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala]: I updated `CacheTableCommand` and `uncacheTable` to support multiparts name (and not resolving the identifier). Please check what you think about the new approach. Thanks.

Comment (imback82) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala]: Also, since it's not resolving to catalogs, we should move it out of `ResolveSessionCatalog`?

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala [hunk 17]
```diff
@@ -442,19 +442,11 @@ class ResolveSessionCatalog(
...
-      val name = if (plan.isDefined) {
-        // CACHE TABLE ... AS SELECT creates a temp view with the input query.
-        // Temp view doesn't belong to any catalog and we shouldn't resolve catalog in the name.
-        tbl
-      } else {
-        parseTempViewOrV1Table(tbl, "CACHE TABLE")
-      }
-      CacheTableCommand(name.asTableIdentifier, plan, isLazy, options)
+    case CacheTable(multipartIdent, plan, isLazy, options) =>
+      CacheTableCommand(multipartIdent, plan, isLazy, options)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:446]: it seems unnecessary to have `CacheTable` now. Should the parser create `CacheTableCommand` directly?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/CachedTableSuite.scala [hunk 18]
```diff
@@ -421,7 +421,7 @@ class CachedTableSuite extends QueryTest with SQLTestUtils with TestHiveSingleto
           activateDatabase(db) {
             sql("REFRESH TABLE default.cachedTable")
             assertCached(
-              sql("SELECT * FROM default.cachedTable"), "`default`.`cachedTable`", DISK_ONLY)
+              sql("SELECT * FROM default.cachedTable"), "cachedTable", DISK_ONLY)
```
Comment (imback82) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/CachedTableSuite.scala:424]: @cloud-fan Looks like we need to resolve catalog/current namespace for this scenario?

Comment (imback82) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/CachedTableSuite.scala:424]: I guess this is OK as long as `spark.catalog.isCached` works fine since it's just a name of cache builder?

Comment (cloud-fan) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/CachedTableSuite.scala:424]: Why is `default.` gone here?

Comment (imback82) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/CachedTableSuite.scala:424]: Because the cache was created with `CACHE TABLE cachedTable OPTIONS('storageLevel' 'DISK_ONLY')`. Before, `cachedTable` was resolved to `default.cachedTable` in `ResolveSessionCatalog`.

Note that it's just a name of cache builder and doesn't affect `spark.catalog.isCached(tableName: String)`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DropTableExec.scala [hunk 19]
```diff
@@ -27,7 +26,6 @@ import org.apache.spark.sql.connector.catalog.{Identifier, Table, TableCatalog}
  * Physical plan node for dropping a table.
  */
 case class DropTableExec(
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DropTableExec.scala:28]: Can we avoid changing it? It's also being changed in https://github.com/apache/spark/pull/30491

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/cache.scala [hunk 20]
```diff
@@ -51,32 +54,35 @@ case class CacheTableCommand(
...
 
-
 case class UncacheTableCommand(
-    tableIdent: TableIdentifier,
+    multipartIdentifier: Seq[String],
     ifExists: Boolean) extends RunnableCommand {
 
   override def run(sparkSession: SparkSession): Seq[Row] = {
-    val tableId = tableIdent.quotedString
-    if (!ifExists || sparkSession.catalog.tableExists(tableId)) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/cache.scala:75]: This reminds me that we should check `CatalogImpl` and see if n-part name is well supported. Seems like `tableExists` is broken...

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/cache.scala:75]: We don't need to fix `CatalogImpl` completely in this PR, but the related ones should be fixed. For `tableExists`:
```
def tableExists(table: String) = tableExists(parseMultiPartName(table))

def tableExists(db:String, table: String) = tableExists(Seq(db, table))

private def tableExists(name: Seq[String]) ...
```

File: sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala [hunk 21]
```diff
@@ -471,8 +471,11 @@ class CatalogImpl(sparkSession: SparkSession) extends Catalog {
    * @since 2.0.0
    */
   override def uncacheTable(tableName: String): Unit = {
-    val tableIdent = sparkSession.sessionState.sqlParser.parseTableIdentifier(tableName)
-    val cascade = !sessionCatalog.isTemporaryTable(tableIdent)
+    import org.apache.spark.sql.connector.catalog.CatalogV2Implicits.MultipartIdentifierHelper
+    val multipartIdentifier =
+      sparkSession.sessionState.sqlParser.parseMultipartIdentifier(tableName)
+    val cascade = (multipartIdentifier.length <= 2) &&
+      !sessionCatalog.isTemporaryTable(multipartIdentifier.asTableIdentifier)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala:478]: can we add an overload of `isTemporaryTable` that takes `Seq[String]`?

Comment (imback82) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala:478]: Looks like the overload already exists as `isTempView`. :)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DropTableExec.scala (hunk 19 above)
Comment (imback82) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DropTableExec.scala:28]: OK, will revert.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/cache.scala (hunk 20 above)
Comment (imback82) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/cache.scala:75]: `CatalogImpl` is for session catalog, so it cannot support v2 tables, no?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/cache.scala:75]: That's a good point, then we can't change `uncacheTable` either...

We need to create a custom `uncacheTable` method in this command as well.

Comment (imback82) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/cache.scala:75]: Ah, I see your point.

`CatalogImpl.cacheTable` already supports v2 tables since it resolves the table name by `sparkSession.table(tableName)`. But, I think `CatalogImpl.cacheTable`, `CatalogImpl.unCacheTable` should only support v1 tables, and `CacheTableCommand` and `UncacheTableCommand` should go thru `sparkSession.sharedState.cacheManager`, right? I will update accordingly.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/cache.scala:75]: yup

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala [hunk 22]
```diff
@@ -2555,11 +2557,15 @@ class DataSourceV2SQLSuite
     }
   }
 
-  private def testNotSupportedV2Command(sqlCommand: String, sqlParams: String): Unit = {
+  private def testNotSupportedV2Command(
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:2560]: unnecessary change. This is minor and let's fix it in your next PR.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/cache.scala [hunk 23]
```diff
@@ -19,26 +19,27 @@ package org.apache.spark.sql.execution.command
...
-import org.apache.spark.sql.{Dataset, Row, SparkSession}
-import org.apache.spark.sql.catalyst.TableIdentifier
+import org.apache.spark.sql.{AnalysisException, DataFrame, Dataset, Row, SparkSession}
 import org.apache.spark.sql.catalyst.plans.QueryPlan
 import org.apache.spark.sql.catalyst.plans.logical.{IgnoreCachedData, LogicalPlan}
 import org.apache.spark.sql.catalyst.util.CaseInsensitiveMap
+import org.apache.spark.sql.connector.catalog.CatalogV2Implicits.MultipartIdentifierHelper
 import org.apache.spark.storage.StorageLevel
 
 case class CacheTableCommand(
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/cache.scala:29]: the next thing we can do is to refactor it using the v2 framework (not adding a v2 version). The benefits are: 1. moving the logical plan to catalyst. 2. resolve the table in the analyzer. e.g.
```
CacheTable(UnresolvedRelation(...), ...)
...
case class CacheTableExec(relation: LogicalPlan) {
  def run() {
     val df  = Dataset.ofRows(spark, relation)
     ....
  }
}
```

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala (hunk 22 above)
Comment (imback82) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:2560]: Ok, will fix.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/cache.scala (hunk 23 above)
Comment (imback82) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/cache.scala:29]: OK, will do.

Comment (imback82) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/cache.scala:29]: One issue I am encountering by moving to the v2 framework (for v2 tables) is the following.

When `CACHE TABLE testcat.tbl` is run, `tbl` is changed from `DataSourceV2Relation` to `DataSourceV2ScanRelation` in `V2ScanRelationPushDown` rule, now that the plan goes thru analyzer, optimizer, etc. But, if I run `spark.table("testcat.tbl")`, the query execution has `tbl` as `DataSourceV2Relation`, thus cache is not applied.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/cache.scala:29]: ah, one solution is to follow `InsertIntoStatement` and do not make the `table` as a child. Then we resolve `UnresolvedRelation` inside `CacheTable` manually in `ResolveTempViews` and other resolution rules.
